File: app/src/main/java/kr/djspi/pipe01/dto/Entry.java
package kr.djspi.pipe01.dto;

import androidx.annotation.NonNull;

import com.google.gson.JsonObject;

import java.io.Serializable;
import java.util.ArrayList;

import lombok.Data;

/**
 * (중요) DTO 클래스는 Proguard 난독화 대상에서 제외시켜야 Json 파싱 과정에서 각종 변수명들이 유지된다.
 * proguard-rules.pro: -keep class kr.djspi.pipe01.dto.** { *; }
 */
@Data
public class Entry implements Serializable {
    public final Spi spi;
    public final SpiType spi_type;
    public final SpiMemo spi_memo;
    public final SpiPhoto spi_photo;
    public final Pipe pipe;
    public final PipeType pipe_type;
    public final PipeShape pipe_shape;
    public final PipePosition pipe_position;
    public final PipePlan pipe_plan;
    public final PipeSupervise pipe_supervise;
    public SpiLocation spi_location;

    public Entry(Spi spi, SpiType spi_type, SpiMemo spi_memo, SpiPhoto spi_photo, Pipe pipe, PipeType pipe_type, PipeShape pipe_shape, PipePosition pipe_position, PipePlan pipe_plan, PipeSupervise pipe_supervise) {
        this.spi = spi;
        this.spi_type = spi_type;
        this.spi_memo = spi_memo;
        this.spi_photo = spi_photo;
        this.pipe = pipe;
        this.pipe_type = pipe_type;
        this.pipe_shape = pipe_shape;
        this.pipe_position = pipe_position;
        this.pipe_plan = pipe_plan;
        this.pipe_supervise = pipe_supervise;
    }

    @SuppressWarnings("SameParameterValue")
    public static JsonObject parseEntry(@NonNull ArrayList entries, int index, String... strings) {
        return ((Entry) entries.get(index)).parseToSingleJsonObject(strings);
    }

    private JsonObject parseToSingleJsonObject(@NonNull String[] strings) {
        JsonObject object = new JsonObject();
        object.addProperty("spi_id", spi.getId());
        object.addProperty("supervise", pipe_supervise.getSupervise());
        object.addProperty("supervise_contact", pipe.getSupervise_contact());
        object.addProperty("construction", pipe.getConstruction());
        object.addProperty("construction_contact", pipe.getConstruction_contact());
        object.addProperty("spi_type", spi_type.getType());
        object.addProperty("pipe", pipe_type.getPipe());
        object.addProperty("shape", pipe_shape.getShape());
        object.addProperty("header", pipe_type.getHeader());
        object.addProperty("spec", pipe_shape.getSpec());
        object.addProperty("unit", pipe_type.getUnit());
        object.addProperty("material", pipe.getMaterial());
        object.addProperty("position", pipe_position.getPosition());
        object.addProperty("direction", pipe_position.getDirection());
        object.addProperty("horizontal", pipe_position.getHorizontal());
        object.addProperty("vertical", pipe_position.getVertical());
        object.addProperty("horizontal_form", strings[0]);
        object.addProperty("vertical_form", strings[1]);
        object.addProperty("depth", pipe.getDepth());
        object.addProperty("spi_memo", spi_memo.getMemo());
        object.addProperty("file_plane", pipe_plan.getFile_plane());
        object.addProperty("file_section", pipe_plan.getFile_section());
        object.addProperty("latitude", spi_location.getLatitude());
        object.addProperty("longitude", spi_location.getLongitude());
        return object;
    }
}

File: app/src/main/java/kr/djspi/pipe01/dto/Entry.kt
package kr.djspi.pipe01.dto

import com.google.gson.JsonObject
import java.io.Serializable

/**
 * (중요) DTO 클래스는 Proguard 난독화 대상에서 제외시켜야 Json 파싱 과정에서 각종 변수명들이 유지된다.
 * proguard-rules.pro: -keep class kr.djspi.pipe01.dto.** { *; }
 */
class Entry(
    val spi: Spi,
    val spi_type: SpiType,
    val spi_memo: SpiMemo,
    val spi_photo: SpiPhoto,
    val pipe: Pipe,
    val pipe_type: PipeType,
    val pipe_shape: PipeShape,
    val pipe_position: PipePosition,
    val pipe_plan: PipePlan,
    val pipe_supervise: PipeSupervise
) : Serializable {

    var spi_location: SpiLocation? = null

    companion object {

        fun parseEntry(entries: ArrayList<*>, index: Int, vararg strings: String?): JsonObject {
            return parseEntrySingle((entries[index] as Entry), strings)
        }

        private fun parseEntrySingle(entry: Entry, strings: Array<out String?>): JsonObject {
            return JsonObject().apply {
                addProperty("spi_id", entry.spi.id)
                addProperty("supervise", entry.pipe_supervise.supervise)
                addProperty("supervise_contact", entry.pipe.supervise_contact)
                addProperty("construction", entry.pipe.construction)
                addProperty("construction_contact", entry.pipe.construction_contact)
                addProperty("spi_type", entry.spi_type.type)
                addProperty("pipe", entry.pipe_type.pipe)
                addProperty("shape", entry.pipe_shape.shape)
                addProperty("header", entry.pipe_type.header)
                addProperty("spec", entry.pipe_shape.spec)
                addProperty("unit", entry.pipe_type.unit)
                addProperty("material", entry.pipe.material)
                addProperty("position", entry.pipe_position.position)
                addProperty("direction", entry.pipe_position.direction)
                addProperty("horizontal", entry.pipe_position.horizontal)
                addProperty("vertical", entry.pipe_position.vertical)
                addProperty("horizontal_form", strings[0])
                addProperty("vertical_form", strings[1])
                addProperty("depth", entry.pipe.depth)
                addProperty("spi_memo", entry.spi_memo.memo)
                addProperty("file_plane", entry.pipe_plan.file_plane)
                addProperty("file_section", entry.pipe_plan.file_section)
                addProperty("latitude", entry.spi_location!!.latitude)
                addProperty("longitude", entry.spi_location!!.longitude)
            }
        }
    }
}
